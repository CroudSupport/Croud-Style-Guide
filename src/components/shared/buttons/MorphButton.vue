<template>
    <button class="ui button" @click="morph()">{{nextOption.name}}</button>
</template>

<script>
/**
* Single Button Component that morphs from one action to another on click.
*
* @example ./croud-morph-button.md
*/

export default {
    name: 'croud-morph-button',
    data() {
        return {
            activeIndex: this.active,
        }
    },
    props: {
        /**
        * The two options to toggle between.
        */
        options: {
            type: Array,
            required: true,
            validator(options) {
                return options.splice(2)
            },
        },
        /**
        * The index of the active action when first rendered.
        */
        active: {
            type: Number,
            default: 0,
        },
    },
    computed: {
        nextOption() {
            return this.options[1 - this.activeIndex]
        },
    },
    methods: {
        morph() {
            this.activeIndex = 1 - this.activeIndex
            this.options[this.activeIndex].method()
        },
    },
}
</script>

<style>

</style>
