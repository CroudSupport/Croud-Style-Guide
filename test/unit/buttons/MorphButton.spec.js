import Vue from 'vue'
import MorphButton from '../../../src/components/shared/buttons/MorphButton'

const Constructor = Vue.extend(MorphButton)

const vm = new Constructor({
    propsData: {
        options: [
            { name: 'close all' },
            { name: 'expand all' },
        ],
        cta: true,
        active: 1,
    },
}).$mount()

describe('Morph Button', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })

    it('should override active index when active prop is passed', () => {
        expect(vm.activeIndex).toBe(vm.active)
    })

    it('should display the correct text on render', () => {
        expect(vm.$el.innerHTML).toBe(vm.options[1 - vm.activeIndex].name)
    })

    it('should toggle text on click', () => {
        vm.$el.click()
        expect(vm.$el.innerHTML).toBe(vm.options[1 - vm.active].name)
        vm.$el.click()
        expect(vm.$el.innerHTML).toBe(vm.options[1 - vm.active].name)
    })
})
