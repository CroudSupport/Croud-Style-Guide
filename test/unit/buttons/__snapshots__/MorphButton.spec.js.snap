// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Morph Button should match the snapshot 1`] = `
<button
  class="ui button"
>
  close all
</button>
`;
